### PR TITLE
Ready to merge

### PR DESCRIPTION
**Pull request by grishasht**
Hope you will enjoy! :)

### CODE REVIEW

File: EpamLab.iml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (diasel2000) [inline EpamLab.iml]: Привет

Comment (grishasht) [reply EpamLab.iml]: Дарова

**Comment by valeriiafatova** [comment]
Commit comments are not sensible

File: EpamLab1.iml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (valeriiafatova) [inline EpamLab1.iml:1]: This files should not be pushed to repository

File: src/main/java/services/Menu.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+import java.util.Scanner;
+import song.Song;
+
+public class Menu {
+    /*Menu interface creation class*/
+    Disc disc = new Disc();
+    DiscTools discTools;
+
+    public Menu() throws IOException {
+        this.discTools = new DiscTools(disc, "src/main/resources/songsList.txt");
```
Comment (valeriiafatova) [inline src/main/java/services/Menu.java]: No logic in constructors!!!!!!!!!!

File: src/main/java/common/Main.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package common;
+
+import java.io.IOException;
+import services.Menu;
+
+public class Main {
+    public static void main(String[] args) throws IOException {
+        new Menu();
```
Comment (valeriiafatova) [inline src/main/java/common/Main.java]: creating an object should not be an entry point. calling method - yes

File: src/main/java/services/CreateIerarchi.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
+package services;
+
+import java.io.IOException;
+import java.rmi.NoSuchObjectException;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import song.Song;
+
+public class CreateIerarchi {
```
Comment (valeriiafatova) [inline src/main/java/services/CreateIerarchi.java]: Class should not be called as Action
Max is "HierarchyCreator". Better HierarchyService

File: src/main/java/services/CreateIerarchi.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    private FileFunctions fileFunctions = FileFunctions.getInstance();
+    private List<Song> songs = new LinkedList();
+
+    public List<Song> getSongs() {
+        return this.songs;
+    }
+
+    private Song getSong(String song) throws IOException {
```
Comment (valeriiafatova) [inline src/main/java/services/CreateIerarchi.java]: Class should not contain AND behavior AND state

File: src/main/java/services/CreateIerarchi.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    public List<Song> addSongToIerarchy(String name, String author, Float duration, String genre) {
+        /*Adds new song to songs list*/
+        songs.add(new Song(name, author, duration, genre));
+        return songs;
+    }
+
+    public List<Song> printSongsList(String filePath) throws IOException {
+        getSongsList(filePath);
+        for(Song song: songs) {
+            System.out.println(song.getName() + " " + song.getAuthor()
```
Comment (valeriiafatova) [inline src/main/java/services/CreateIerarchi.java]: Advise to use or stringbuilder or system out print f

File: src/main/java/services/DiscTools.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    public DiscTools(Disc disc, String filePath) {
+        this.disc = disc;
+        this.filePath = filePath;
+    }
+
+    public String getFilePath() {
+        return this.filePath;
+    }
+
+    public void writeSongsOnDisc() throws IOException {
```
Comment (valeriiafatova) [inline src/main/java/services/DiscTools.java:27]: Move logic to Service

File: src/main/java/services/DiscTools.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        /*Records songs from file on disc*/
+        disc.writeOnDisc(this.ierarchi.getSongsList(this.filePath));
+    }
+
+    public List<Song> addSongOnDisc(String name, String author, Float duration, String genre) throws IOException {
+        /*Adds new users song to songs list on the disc*/
```
Comment (valeriiafatova) [inline src/main/java/services/DiscTools.java]: Add comment before method, not IN

File: src/main/java/services/Menu.java (hunk 3 above)
Comment (valeriiafatova) [inline src/main/java/services/Menu.java]: Who catch IOException?

File: src/main/java/song/Song.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+    public void setDuration(Float duration) {
+        this.duration = duration;
+    }
+
+    public void setGenre(String genre) {
+        this.genre = genre;
+    }
+
+    public boolean equals(Object o) {
```
Comment (valeriiafatova) [inline src/main/java/song/Song.java:51]: Where check for null?

File: target/classes/songsList.txt [hunk 11]
```diff
@@ -0,0 +1,17 @@
+Candy shop-50 Cent-4.15-Rap-
```
Comment (valeriiafatova) [inline target/classes/songsList.txt]: target folder should not be commited to repo

File: src/main/java/services/Sort.java [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+                .comparing(Song::getName)).collect(Collectors.toList());
+    }
+
+    public static List<Song> sortByAuthor(Disc disc) {
+        return disc.getSongs().stream().sorted(Comparator
+                .comparing(Song::getAuthor)).collect(Collectors.toList());
+    }
+
+    public static List<Song> sortByDuration(Disc disc) {
+        return disc.getSongs().stream().sorted(Comparator
```
Comment (valeriiafatova) [inline src/main/java/services/Sort.java]: every stream method should start from new line

File: src/main/java/services/Sort.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
+package services;
+
+import disc.Disc;
+import java.util.Comparator;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import song.Song;
+
+public class Sort {
```
Comment (valeriiafatova) [inline src/main/java/services/Sort.java:10]: Sort looks more like service. Rename to SortService, please

File: src/main/java/services/Menu.java [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+        this.printMenu();
+        this.initDisc(discTools);
+    }
+
+    public void printMenu() throws IOException {
+        /*Menu interface*/
+        Boolean f = true;
+        System.out.println("Welcome in my studio!\n" +
```
Comment (valeriiafatova) [inline src/main/java/services/Menu.java]: Use StringBuilder

File: src/main/java/services/Menu.java [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+                "Record songs on cd - 1\n" +
+                "Calculate general songs duration - 2\n" +
+                "Sort contains on disc - 3\n" +
+                "Find composition on your disc - 4\n" +
+                "Remove composition from disc - 5\n" +
+                "Look at disc contains - 6\n" +
+                "Fill disc with existing compositions - 7\n" +
+                "Exit from program - 0\n");
+
+        for(; f; System.out.println()) {
```
Comment (valeriiafatova) [inline src/main/java/services/Menu.java]: what 'f' means?

File: src/main/java/services/Menu.java [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+            Scanner scanner = new Scanner(System.in);
+            System.out.print("So what is your choice: ");
+            Integer userChoice = scanner.nextInt();
+            switch(userChoice) {
+                case 0: f = false; break;
+                case 1: addSongOnDisc(discTools); break;
```
Comment (valeriiafatova) [inline src/main/java/services/Menu.java]: break should start from new line

File: src/main/java/services/FileFunctions.java [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.Iterator;
+import java.util.List;
+import song.Song;
+
+public class FileFunctions {
```
Comment (valeriiafatova) [inline src/main/java/services/FileFunctions.java]: Perhaps FileService would Sound better?

File: src/main/java/services/FileFunctions.java [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+
+    private static class SingletonHolder {
+        private static final FileFunctions INSTANCE = new FileFunctions();
+
+        private SingletonHolder() {
+        }
+    }
+
+    public void openFile(String filePath) throws IOException {
+        buffReader = new BufferedReader(new FileReader(filePath));
```
Comment (valeriiafatova) [inline src/main/java/services/FileFunctions.java]: methods depends from each other. If you call 'read' before 'open' you ll get NullPointer.

File: src/main/java/services/FileFunctions.java [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.Iterator;
+import java.util.List;
+import song.Song;
+
+public class FileFunctions {
+    public BufferedReader buffReader;
+
+    public static FileFunctions getInstance() {
```
Comment (valeriiafatova) [inline src/main/java/services/FileFunctions.java]: Why singleton?

File: src/main/java/services/DiscTools.java [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+                        + " -- " + song.getDuration() + " -- " + song.getGenre());
+    }
+
+    public Float getFullDuration() {
+        /*Returns general songs duration from disc*/
+        List<Song> songs = disc.getSongs();
+        Float fullDuration = 0.0F;
+        Song song;
+        if (songs != null) {
+            for(Iterator iter = songs.iterator(); iter.hasNext(); fullDuration = fullDuration + song.getDuration()) {
```
Comment (valeriiafatova) [inline src/main/java/services/DiscTools.java]: better to use Streams

File: src/main/java/services/DiscTools.java [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+                song = (Song)iter.next();
+            }
+        }
+
+        return fullDuration;
+    }
+
+    public List<Song> songsByDuration(Integer from, Integer to) {
+        /*Returns list of songs from entered by user range*/
+        return disc.getSongs().stream().filter(song -> song.getDuration() <=
```
Comment (valeriiafatova) [inline src/main/java/services/DiscTools.java]: each stream from new line

File: src/main/java/services/DiscTools.java [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+        /*Removes song by name from disc*/
+        disc.removeFromDisc(songName);
+        fileFunctions.writeInFile(filePath, disc.getSongs());
+        return disc.getSongs();
+    }
+
+    public void printSongs(List<Song> songs) throws IOException {
+        if (songs != null)
+            for(Song song: songs)
+                System.out.println(song.getName() + " -- " + song.getAuthor()
```
Comment (valeriiafatova) [inline src/main/java/services/DiscTools.java]: Better to use StringBuilde

File: src/main/java/services/DiscTools.java [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+        fileFunctions.writeInFile(filePath, ierarchi.getSongs());
+        writeSongsOnDisc();
+        return disc.getSongs();
+    }
+
+    public List<Song> removeSongFromDisc(String songName) throws IOException {
+        /*Removes song by name from disc*/
+        disc.removeFromDisc(songName);
+        fileFunctions.writeInFile(filePath, disc.getSongs());
+        return disc.getSongs();
```
Comment (valeriiafatova) [inline src/main/java/services/DiscTools.java:49]: what do you return in 'delete' method? Does removing requires also returning other songs?

File: src/main/java/services/CreateIerarchi.java [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+                        throw new NoSuchObjectException("No such object");
+                }
+        }
+        return songInit;
+    }
+
+    public List<Song> getSongsList(String filePath) throws IOException {
+        /*Reads songs from file and pushes it into songs list*/
+        fileFunctions.openFile(filePath);
+        songs.clear();
```
Comment (valeriiafatova) [inline src/main/java/services/CreateIerarchi.java]: why you call clear after 'open' ?

File: src/main/java/services/CreateIerarchi.java (hunk 6 above)
Comment (valeriiafatova) [reply src/main/java/services/CreateIerarchi.java]: Move to service

File: src/main/java/disc/Disc.java [hunk 25]
```diff
@@ -0,0 +1,25 @@
...
+    private List<Song> songs;
+
+    public Disc() {
+    }
+
+    public void writeOnDisc(List<Song> songs) {
+        this.songs = songs;
+    }
+
+    public void removeFromDisc(String songName) {
```
Comment (valeriiafatova) [inline src/main/java/disc/Disc.java]: Move to Service

**Comment by grishasht** [comment]
> Commit comments are not sensible
Sorry, I hadn't wanted to make any commits then. But unexpectedly happened story with git and my project which I'd told in telegram and I was forced. And, so, my commits have no important changes which I could notice in their names, just deleted comments from files

File: src/main/java/services/CreateIerarchi.java (hunk 24 above)
Comment (grishasht) [reply src/main/java/services/CreateIerarchi.java]: Cause every time I call this method, it writes up a new list to already existing. So I clear it and write new instead

File: src/main/java/services/CreateIerarchi.java (hunk 7 above)
Comment (grishasht) [reply src/main/java/services/CreateIerarchi.java]: Deleted this method, it's redundant

File: src/main/java/services/DiscTools.java (hunk 8 above)
Comment (grishasht) [reply src/main/java/services/DiscTools.java:27]: But it is a service class O_O

File: src/main/java/services/DiscTools.java (hunk 9 above)
Comment (grishasht) [reply src/main/java/services/DiscTools.java]: Fixed

File: src/main/java/common/Main.java (hunk 4 above)
Comment (grishasht) [reply src/main/java/common/Main.java]: Fixed

File: EpamLab1.iml (hunk 2 above)
Comment (grishasht) [reply EpamLab1.iml:1]: Fixed

File: src/main/java/disc/Disc.java (hunk 25 above)
Comment (grishasht) [reply src/main/java/disc/Disc.java]: Fixed

File: src/main/java/services/CreateIerarchi.java (hunk 5 above)
Comment (grishasht) [reply src/main/java/services/CreateIerarchi.java]: Fixed

File: src/main/java/services/CreateIerarchi.java (hunk 6 above)
Comment (grishasht) [reply src/main/java/services/CreateIerarchi.java]: Fixed

File: src/main/java/services/DiscTools.java (hunk 23 above)
Comment (grishasht) [reply src/main/java/services/DiscTools.java:49]: Yeap, it's necessary for Unit tests

File: src/main/java/services/DiscTools.java (hunk 22 above)
Comment (grishasht) [reply src/main/java/services/DiscTools.java]: Fixed

File: src/main/java/services/DiscTools.java (hunk 20 above)
Comment (grishasht) [reply src/main/java/services/DiscTools.java]: Fixed

File: src/main/java/services/DiscTools.java (hunk 21 above)
Comment (grishasht) [reply src/main/java/services/DiscTools.java]: Fixed

File: src/main/java/services/FileFunctions.java (hunk 17 above)
Comment (grishasht) [reply src/main/java/services/FileFunctions.java]: Fixed

File: src/main/java/services/FileFunctions.java (hunk 19 above)
Comment (grishasht) [reply src/main/java/services/FileFunctions.java]: Fixed

File: src/main/java/services/FileFunctions.java (hunk 18 above)
Comment (grishasht) [reply src/main/java/services/FileFunctions.java]: Fixed

File: src/main/java/services/Menu.java (hunk 3 above)
Comment (grishasht) [reply src/main/java/services/Menu.java]: Fixed

Comment (grishasht) [reply src/main/java/services/Menu.java]: Fixed

File: src/main/java/services/Menu.java (hunk 14 above)
Comment (grishasht) [reply src/main/java/services/Menu.java]: Fixed

File: src/main/java/services/Menu.java (hunk 15 above)
Comment (grishasht) [reply src/main/java/services/Menu.java]: Fixed

File: src/main/java/services/Menu.java (hunk 16 above)
Comment (grishasht) [reply src/main/java/services/Menu.java]: Fixed

File: src/main/java/services/Sort.java (hunk 13 above)
Comment (grishasht) [reply src/main/java/services/Sort.java:10]: Fixed

File: src/main/java/services/Sort.java (hunk 12 above)
Comment (grishasht) [reply src/main/java/services/Sort.java]: Fixed

File: src/main/java/song/Song.java (hunk 10 above)
Comment (grishasht) [reply src/main/java/song/Song.java:51]: Fixed. It is in the condition below. Created by Intellij

File: target/classes/songsList.txt (hunk 11 above)
Comment (grishasht) [reply target/classes/songsList.txt]: Fixed

File: src/main/java/services/DiscTools.java [hunk 26]
```diff
@@ -0,0 +1,93 @@
...
+            songs.stream()
+                    .mapToDouble(Song::getDuration)
+                    .sum();
+        }
+        return fullDuration;
+    }
+
+    /*Returns list of songs from entered by user range*/
+    public List<Song> songsByDuration(Integer from, Integer to) {
+        return disc.getSongs().stream().filter(song -> song.getDuration() <= (float) to
```
Comment (valeriiafatova) [inline src/main/java/services/DiscTools.java:72]: formatting

File: src/main/java/services/FileServices.java [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+
+public class FileServices {
+    public BufferedReader buffReader;
+
+    public void openFile (String filePath) throws IOException {
+        buffReader = new BufferedReader(new FileReader(filePath));
+    }
+
+    public String readFromFile (String filePath) throws IOException {
+        if (!buffReader.ready()) openFile(filePath);
```
Comment (valeriiafatova) [inline src/main/java/services/FileServices.java:21]: from new line

File: src/main/java/services/HierarchyCreator.java [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+
+    public List<Song> getSongs() {
+        return this.songs;
+    }
+
+    /*Parses gotten string and returns gotten song*/
+    private Song getSong(String song) throws IOException {
+        Song songInit = new Song();
+        String buff = "";
+        int k = 0;
```
Comment (valeriiafatova) [inline src/main/java/services/HierarchyCreator.java:21]: what 'k' means?

File: src/main/java/services/HierarchyCreator.java [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+    private List<Song> songs = new LinkedList();
+
+    public List<Song> getSongs() {
+        return this.songs;
+    }
+
+    /*Parses gotten string and returns gotten song*/
+    private Song getSong(String song) throws IOException {
+        Song songInit = new Song();
+        String buff = "";
```
Comment (valeriiafatova) [inline src/main/java/services/HierarchyCreator.java:20]: Change to String Builder

File: src/main/java/services/HierarchyCreator.java [hunk 30]
```diff
@@ -0,0 +1,60 @@
...
+        int k = 0;
+
+        for(int i = 0; i < song.length(); ++i) {
+            char ch = song.charAt(i);
+            if (ch != '-') {
```
Comment (valeriiafatova) [inline src/main/java/services/HierarchyCreator.java:25]: do wth pattern

File: src/main/java/services/Menu.java [hunk 31]
```diff
@@ -0,0 +1,112 @@
...
+        FileFunctions fileFunctions = FileFunctions.getInstance();
+        List<Song> sortedSongs = discTools.sortSongs(scanner.next());
+        fileFunctions.writeInFile(discTools.getFilePath(), sortedSongs);
+        discTools.printSongs(sortedSongs);
+        System.out.println("Successfully sorted! Init the disc");
+    }
+
+    /*Finds composition from entered by user range*/
+    private void findCompositionOnDisc(DiscTools discTools) throws IOException {
+        Scanner scanner = new Scanner(System.in);
```
Comment (valeriiafatova) [inline src/main/java/services/Menu.java:90]: Move Scanner to separate place

**Pull request merged**